flask/app/tissue_samples.py

```diff
@@ -0,0 +1,179 @@
+
+
+@app.route("/api/tissue_samples", methods=["POST"])
+@login_required
+@check_admin
+def create_tissue_sample():
+    if not request.json:
+        return "Request body must be JSON", 400
+
+    tissue_sample_type = request.json.get("tissue_sample_type")
+    if not tissue_sample_type:
+        return "A tissue sample type must be provided", 400
+
+    participant_id = request.json.get("participant_id")
+    if not participant_id:
+        return "A participant id must be provided", 400
+
+    models.Participant.query.filter_by(participant_id=participant_id).first_or_404()
+
+    enum_error = enum_validate(models.TissueSample, request.json, editable_columns)
+
+    if enum_error:
+        return enum_error, 400
+
+    try:
+        created_by = updated_by = current_user.user_id
+    except:  # LOGIN DISABLED
+        created_by = updated_by = 1
+
+    tissue_sample = models.TissueSample(
+        **{
+            "participant_id": participant_id,
+            "extraction_date": request.json.get("extraction_date"),
+            "tissue_sample_type": tissue_sample_type,
+            "tissue_processing": request.json.get("tissue_processing"),
+            "notes": request.json.get("notes"),
+            "created_by": created_by,
+            "updated_by": updated_by,
+        }
+    )
+    try:
+        db.session.add(tissue_sample)
+        db.session.commit()
+        ts_id = tissue_sample.tissue_sample_id
+        location_header = "/api/tissue_samples/{}".format(ts_id)
+        return jsonify(tissue_sample), 201, {"location": location_header}
+    except:
+        db.session.rollback()
+        return "Server error", 500
+
+
+@app.route("/api/tissue_samples/<int:id>", methods=["PATCH"])
+@login_required
+@check_admin
+def update_tissue_sample(id: int):
+    if not request.json:
+        return "Request body must be JSON", 415
+
+    if app.config.get("LOGIN_DISABLED") or current_user.is_admin:
+        user_id = request.args.get("user")
+    else:
+        user_id = current_user.user_id
+
+    if user_id:
+        tissue_sample = (
+            models.TissueSample.query.filter_by(tissue_sample_id=id)
+            .options(contains_eager(models.TissueSample.datasets))
+            .join(models.Dataset)
+            .join(
+                models.groups_datasets_table,
+                models.Dataset.dataset_id
+                == models.groups_datasets_table.columns.dataset_id,
+            )
+            .join(
+                models.users_groups_table,
+                models.groups_datasets_table.columns.group_id
+                == models.users_groups_table.columns.group_id,
+            )
+            .filter(models.users_groups_table.columns.user_id == user_id)
+            .one_or_none()
+        )
+    else:
+        tissue_sample = (
+            models.TissueSample.query.filter_by(tissue_sample_id=id)
+            .options(joinedload(models.TissueSample.datasets))
+            .one_or_none()
+        )
+
```

Comment:
if the tissue_sample_id doesn't exist, the endpoint returns an unrelated error (because of something else failing downstream as a result of `tissue_sample` not existing?). I think there should be some kind of validation for whether the tissue_sample_id exists or whether the tissue_sample object exists

Reply:
resolved in [4068440](https://github.com/ccmbioinfo/ST2020/pull/241/commits/40684408eedeeb532e403ab8203e8cc65edb1aef)